zio-http/shared/src/main/scala/zio/http/URL.scala

```diff
@@ -279,26 +280,35 @@ final case class URL(
 }
 
 object URL {
-  def empty: URL = URL(Path.empty)
+  val empty: URL = URL(path = Path.empty)
+
+  /**
+   * Proper implementation of [[java.net.MalformedURLException]] which, unlike
+   * the Java version, can propagate the cause.
+   */
+  final case class MalformedURLException private[http] (rawUrl: String, cause: Option[Throwable])
```

Comment:
I make the constructor private so people can't build a `MalformedURLException` themselves as it's IMO internals.

WDYT?

Reply:
I don't think we need this to be public at all. Wdut about something like this:

```scala
  final private case class Err(rawUrl: String, cause: Throwable) extends MalformedURLException {
    override def getMessage: String  = s"""Invalid URL: "$rawUrl""""
    override def getCause: Throwable = cause
  }
```

Reply:
It needs to be public I think as it's part of the result of the public function `URL.decode`, no? 🤔

Reply:
The return of `decode` can be `Either[java.net.MalformedURLException, URL]` (since `Err extends java.net.MalformedURLException`)

Reply:
@kyri-petrou I'm sorry I'm not sure to understand the benefit of your suggestion 🤔
What if the here wants to have access to the `cause`?

Reply:
No the user can still access it because the `getCause` method is defined on `Throwable`. My suggestion is so that we don't have to return a custom error (which might cause issues with binary compatibility in the future if we want to extend it)

Reply:
https://scastie.scala-lang.org/3TFy9PhyTPqomnoOcfXl4Q

Reply:
Done